rubix-tests/src/test/java/com/qubole/rubix/tests/TestGenerationNumber.java

```diff
@@ -0,0 +1,279 @@
+  }
+
+  @Test
+  void testReadWithInvalidations() throws Exception {
+    testReadWithInvalidationHelper();
+    CacheConfig.setIsParallelWarmupEnabled(conf, true);
+    restartServer();
+    testReadWithInvalidationHelper();
+  }
+
+  void testReadWithInvalidationHelper() throws TException, InterruptedException {
+    final int readOffset = 0, readLength = 2000;
+    RetryingPooledBookkeeperClient client = getBookKeeperClient();
+    Thread invalidateRequest = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        for (int i = 1; i < 15; i++)
+        {
+          try {
+            client.invalidateFileMetadata(backendPath.toString());
+            Thread.sleep(1000);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    });
+    Thread readrequest = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        for (int i = 1; i < 200; i++)
+        {
+          try {
+            byte[] buffer = new byte[readLength];
+            inputStream.seek(0);
+            int readSize = inputStream.read(buffer, 0, readLength);
+            assertTrue(readSize == readLength, "Wrong amount of data read " + readSize + " was expecting " + readLength);
+            Thread.sleep(100);
+          }
+          catch (Exception e)
+          {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    });
+    invalidateRequest.start();
+    readrequest.start();
+    readrequest.join();
+    invalidateRequest.join();
+  }
+
+  @Test
+  void testGenerationNumberIncrement() throws TException, IOException
+  {
+    RetryingPooledBookkeeperClient client = getBookKeeperClient();
+    final int readOffset = 0, readLength = 2000;
+    client.readData(backendPath.toString(), readOffset, readLength, TEST_FILE_LENGTH, TEST_LAST_MODIFIED, ClusterType.TEST_CLUSTER_MANAGER.ordinal());
+    client.invalidateFileMetadata(backendPath.toString());
+    CacheStatusRequest request = new CacheStatusRequest(backendPath.toString(), TEST_FILE_LENGTH, TEST_LAST_MODIFIED,
+        0, 20).setClusterType(ClusterType.TEST_CLUSTER_MANAGER.ordinal());
+    int generationNumber = client.getCacheStatus(request).getGenerationNumber();
+    assertTrue(generationNumber == UNKONWN_GENERATION_NUMBER + 2,
+        "generation number is reported: " + generationNumber +
+        "but expected: " + UNKONWN_GENERATION_NUMBER + 2);
+  }
+
+  @Test
+  void testGenerationNumberDuringStartUp() throws Exception
+  {
+    int generationNumber = 5;
+    // create files till generation Number = 5
+    creatLocalFilesOnCache(generationNumber);
+    boolean missingFile = false;
+    testGenerationNumberDuringStartUpHelper(generationNumber, missingFile);
+    missingFile = true;
+    restartServer();
+    creatLocalFilesOnCache(generationNumber);
+    new File(CacheUtil.getMetadataFilePath(backendPath.toString(), conf, generationNumber)).delete();
```

Comment:
There are some cases still left. Cover all the cases covered in FileMetadata constructor to ensure no regressions in future.